org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/MapExtensions.java

```diff
@@ -110,4 +122,484 @@ public boolean apply(Map.Entry<K, V> input) {
+	 * @param <V> - type of the map values.
+	 * @param map the map to update.
+	 * @param keys the keys of the pairs to remove.
+	 * @return the map with the content of the map except the pairs.
+	 * @since 2.14
+	 */
+	@Pure
+	public static <K, V> Map<K, V> operator_minus(Map<K, V> map, final Iterable<?> keys) {
+		return Maps.filterKeys(map, new Predicate<K>() {
+			@Override
+			public boolean apply(K input) {
+				return !Iterables.contains(keys, input);
+			}
+		});
+	}
+
+	/**
+	 * Merge the given maps.
+	 *
+	 * <p>
+	 * The replied map is a view on the given two maps.
+	 * If a key exsits in the two maps, the replied value is the value of the right operand.
+	 * </p>
+	 *
+	 * <p>The replied map is unmodifiable.
+	 *
+	 * @param <K> - type of the map keys.
+	 * @param <V> - type of the map values.
+	 * @param left the left map.
+	 * @param right the right map.
+	 * @return a map with the merged contents from the two maps.
+	 * @since 2.14
+	 */
+	public static <K, V> Map<K, V> union(Map<? extends K, ? extends V> left, Map<? extends K, ? extends V> right) {
+		return new MergingMap<K, V>(left, right);
+	}
+
+	private static class MergingMap<K, V> implements Map<K, V> {
+
+		private final Map<? extends K, ? extends V> left;
+
+		private final Map<? extends K, ? extends V> filteredLeft;
+
+		private final Map<? extends K, ? extends V> right;
+
+		MergingMap(Map<? extends K, ? extends V> left, Map<? extends K, ? extends V> right) {
+			this.left = left;
+			this.right = right;
+			this.filteredLeft = Maps.filterKeys(this.left, new Predicate<K>() {
+				@SuppressWarnings("synthetic-access")
+				@Override
+				public boolean apply(K input) {
+					return !MergingMap.this.right.containsKey(input);
+				}
+
+			});
+		}
+
+		@Override
+		public int size() {
+			return this.right.size() + this.filteredLeft.size();
+		}
+
+		@Override
+		public boolean isEmpty() {
+			return this.right.isEmpty() && this.left.isEmpty();
+		}
+
+		@Override
+		public boolean containsKey(Object key) {
+			return this.right.containsKey(key) || this.left.containsKey(key);
+		}
+
+		@Override
+		public boolean containsValue(Object value) {
+			if (this.right.containsValue(value)) {
+				return true;
+			}
+			return this.filteredLeft.containsValue(value);
+		}
+
+		@Override
+		public V get(Object key) {
+			if (this.right.containsKey(key)) {
+				return this.right.get(key);
+			}
+			return this.left.get(key);
+		}
+
+		@Override
+		public V put(K key, V value) {
+			throw new UnsupportedOperationException();
+		}
+
+		@Override
+		public V remove(Object key) {
+			throw new UnsupportedOperationException();
+		}
+
+		@Override
+		public void putAll(Map<? extends K, ? extends V> map) {
+			throw new UnsupportedOperationException();
+		}
+
+		@Override
+		public void clear() {
+			throw new UnsupportedOperationException();
+		}
+
+		@Override
+		public Set<K> keySet() {
+			return Sets.union(this.right.keySet(), this.filteredLeft.keySet());
```

Comment:
Can be a union of this.right.keySet and this.left.keySet

Reply:
Indeed, at this place of the code, `filteredLeft` could be safely replaced by `left`. I will do.